- Description: qeth: incorrect handling of ipa_takeover/enable attribute
  Symptom:     The mentioned attribute has no effect on enabling takeover
               at card setup time.
  Problem:     Takeover enabled/disabled flags are always taken from a
               static array, no matter if ipa_takeover is enabled or not.
  Solution:    Remove static initialization of ipa_takeover. Introduced
               function qeth_init_func_level.
  Problem-ID:  --
  Archs:       s390-31, s390-64

--- linux-2.5/drivers/s390/net/qeth_main.c	14 May 2004 08:16:06 -0000	1.77.2.8
+++ linux-2.5/drivers/s390/net/qeth_main.c	14 May 2004 08:30:14 -0000	1.77.2.9
@@ -1,6 +1,6 @@
 /*
  * 
- * linux/drivers/s390/net/qeth_main.c ($Revision: 1.77.2.8 $)
+ * linux/drivers/s390/net/qeth_main.c ($Revision: 1.77.2.9 $)
  *
  * Linux on zSeries OSA Express and HiperSockets support
  *
@@ -12,7 +12,7 @@
  *			  Frank Pavlic (pavlic@de.ibm.com) and
  *		 	  Thomas Spatzier <tspat@de.ibm.com>
  *
- *    $Revision: 1.77.2.8 $	 $Date: 2004/05/14 08:16:06 $
+ *    $Revision: 1.77.2.9 $	 $Date: 2004/05/14 08:30:14 $
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -78,7 +78,7 @@
 #include "qeth_mpc.h"
 #include "qeth_fs.h"
 
-#define VERSION_QETH_C "$Revision: 1.77.2.8 $"
+#define VERSION_QETH_C "$Revision: 1.77.2.9 $"
 static const char *version = "qeth S/390 OSA-Express driver";
 
 /**
@@ -933,7 +933,6 @@
 	card->options.checksum_type = QETH_CHECKSUM_DEFAULT;
 	card->options.broadcast_mode = QETH_TR_BROADCAST_ALLRINGS;
 	card->options.macaddr_mode = QETH_TR_MACADDR_NONCANONICAL;
-	card->options.enable_takeover = 1;
 	card->options.fake_broadcast = 0;
 	card->options.add_hhlen = DEFAULT_ADD_HHLEN;
 	card->options.fake_ll = 0;
@@ -994,10 +993,6 @@
 		if ((CARD_RDEV(card)->id.dev_type == known_devices[i][2]) &&
 		    (CARD_RDEV(card)->id.dev_model == known_devices[i][3])) {
 			card->info.type = known_devices[i][4];
-			if (card->options.enable_takeover)
-				card->info.func_level = known_devices[i][6];
-			else
-				card->info.func_level = known_devices[i][7];
 			card->qdio.no_out_queues = known_devices[i][8];
 			card->info.is_multicast_different = known_devices[i][9];
 			return 0;
@@ -4825,6 +4820,26 @@
 	return 0;
 }
 
+static void
+qeth_init_func_level(struct qeth_card *card)
+{
+	if (card->ipato.enabled) {
+		if (card->info.type == QETH_CARD_TYPE_IQD)
+				card->info.func_level =
+					QETH_IDX_FUNC_LEVEL_IQD_ENA_IPAT;
+		else
+				card->info.func_level =
+					QETH_IDX_FUNC_LEVEL_OSAE_ENA_IPAT;
+	} else {
+		if (card->info.type == QETH_CARD_TYPE_IQD)
+			card->info.func_level =
+				QETH_IDX_FUNC_LEVEL_IQD_DIS_IPAT;
+		else
+			card->info.func_level =
+				QETH_IDX_FUNC_LEVEL_OSAE_DIS_IPAT;
+	}
+}
+
 /**
  * hardsetup card, initialize MPC and QDIO stuff 
  */ 
@@ -4862,6 +4877,7 @@
 		return rc;
 	}
 	qeth_init_tokens(card);
+	qeth_init_func_level(card);
 	rc = qeth_idx_activate_channel(&card->read, qeth_idx_read_cb);
 	if (rc == -ERESTARTSYS) {
 		QETH_DBF_TEXT(setup, 2, "break2");

--- linux-2.5/drivers/s390/net/qeth_sys.c	14 May 2004 07:27:38 -0000	1.19.2.4
+++ linux-2.5/drivers/s390/net/qeth_sys.c	14 May 2004 08:31:13 -0000	1.19.2.5
@@ -1,6 +1,6 @@
 /*
  *
- * linux/drivers/s390/net/qeth_sys.c ($Revision: 1.19.2.4 $)
+ * linux/drivers/s390/net/qeth_sys.c ($Revision: 1.19.2.5 $)
  *
  * Linux on zSeries OSA Express and HiperSockets support
  * This file contains code related to sysfs.
@@ -20,7 +20,7 @@
 #include "qeth_mpc.h"
 #include "qeth_fs.h"
 
-const char *VERSION_QETH_SYS_C = "$Revision: 1.19.2.4 $";
+const char *VERSION_QETH_SYS_C = "$Revision: 1.19.2.5 $";
 
 /*****************************************************************************/
 /*                                                                           */
@@ -738,6 +738,10 @@
 
 	if (!card)
 		return -EINVAL;
+
+	if ((card->state != CARD_STATE_DOWN) &&
+	    (card->state != CARD_STATE_RECOVER))
+		return -EPERM;
 
 	tmp = strsep((char **) &buf, "\n");
 	if (!strcmp(tmp, "toggle")){

--- linux-2.5/drivers/s390/net/qeth.h	14 May 2004 08:16:06 -0000	1.98.2.4
+++ linux-2.5/drivers/s390/net/qeth.h	14 May 2004 08:31:13 -0000	1.98.2.5
@@ -23,7 +23,7 @@
 
 #include "qeth_mpc.h"
 
-#define VERSION_QETH_H 		"$Revision: 1.98.2.4 $"
+#define VERSION_QETH_H 		"$Revision: 1.98.2.5 $"
 
 #ifdef CONFIG_QETH_IPV6
 #define QETH_VERSION_IPV6 	":IPv6"
@@ -648,7 +648,6 @@
 	enum qeth_checksum_types checksum_type;
 	int broadcast_mode;
 	int macaddr_mode;
-	int enable_takeover;
 	int fake_broadcast;
 	int add_hhlen;
 	int fake_ll;

